src/tuner/nccl_ofi_tuner.c

```diff
@@ -0,0 +1,125 @@
+#include <stdlib.h>
+#include "nccl-headers/nvidia/tuner.h"
+#include "nccl_ofi_tuner.h"
+#include "nccl_ofi_log.h"
+
+ncclResult_t nccl_ofi_tuner_init(size_t nRanks, size_t nNodes, ncclDebugLogger_t logFunction)
+{
+	ofi_log_function = logFunction;
+
+	/*
+	 * NCCL parses these variables and applies user filters inside its
+	 * current tuner logic. Ideally, this should be done regardless of the
+	 * use of NCCL's internal tuner or an external tuner plugin. For the
+	 * time being, given the external tuner is an opt-in, detect if a user
+	 * has set one of them and bail when an external tuner is loaded.
+	 */
+	if (getenv("NCCL_ALGO") || getenv("NCCL_PROTO")) {
+		NCCL_OFI_WARN("The tuner plugin can not be loaded when explicitly choosing an algorithm or protocol with NCCL_ALGO/NCCL_PROTO");
+		return ncclInvalidUsage;
+	}
+
+	struct nccl_ofi_tuner_model_params params = {
+		.net_lat = ofi_nccl_tuner_net_latency(),
+		.internode_bw = NCCL_OFI_TUNER_INTERNODE_BW,
+		.intranode_bw = NCCL_OFI_TUNER_INTRANODE_BW,
+		.num_rails = NCCL_OFI_TUNER_NET_NUM_RAILS
+	};
+
+	/*
+	 * The tuner API is missing a mechanism to pass around context after
+	 * initialization. We currently do not maintain any
+	 * communicator-specific state in the context, but if we do, multiple
+	 * invocations would overwrite data in the global context which is not
+	 * correct. For now, detecting the re-init and warning. This should turn
+	 * into a failure if we include communicator-specific state in the
+	 * context.
+	 */ 
+	if (nccl_ofi_tuner_ctx) {
+		NCCL_OFI_WARN("Context already initialized for communicator\n");
```

Comment:
why should this be a warn?  This is also still racy.  Why not have a static initialized pthread mutex and take that lock around this initialization?

The `tuner_model_costs()` call should also be in the initialize once check.

I'd write this so that if the context ends up being a return value, there's minimal changes to this code.  We're not sure which way Nvidia is going to go, so factor accordingly.